### PR TITLE
Add a new gitian descriptor to enable parallel gitian builds

### PR DESCRIPTION
**Pull request by superbaud**
To try it out, run `git clone https://github.com/superbaud/zcash-gitian -b parallel-build` (or do `git switch parallel-build` in your `zcash-gitian` repo), and set the `.env` therein to:
```
GPG_KEY_ID=16460E1BC0E0507E
GPG_KEY_NAME=sasha
ZCASH_GIT_REPO_URL=https://github.com/superbaud/zcash
ZCASH_VERSION=parallel-gitian-descriptor

GITIAN_BUILDER_URL=https://github.com/devrandom/gitian-builder
GITIAN_BUILDER_VERSION=master
``` 

Until this PR is merged, you will need to target a `ZCASH_VERSION` which has the new parallel gitian descriptor included.

Run `vagrant destroy zcash-build && vagrant up zcash-build`, and once the Vagrant VM is up, run `./gitian-parallel-build.sh`. `./gitian-build.sh` works as previously.

To be sure that we still have reproducibility, it would be good if someone else ran both `./gitian-build.sh`  and `./gitian-parallel-build.sh` (**in clean instances of the Vagrant VM, be sure you run** `vagrant destroy zcash-build` **before each time you run either command**) targeting the above repo URL and branch name to verify that they get the same hashes as I do.

### CODE REVIEW

**Comment by superbaud** [comment]
Bullseye hashes with the parallel gitian descriptor, using `gitian-parallel-build.sh` on a 16-core GCP VM:

```
vagrant@zcash-build:~/gitian.sigs$ head parallel-gitian-descriptor_bullseye/sasha/zcash-5.2.0-rc1-build.assert
--- !!omap
- out_manifest: |
    74ecf49e8edbe694216ecca2b2623df7520a59cdd26c68bf2bc1bb051630cced  src/zcash-5.2.0.tar.gz
    e71f25288c2fd9e548732bd2e2235f9029ff014cfbcfb4d35a7e1c63446f89e7  zcash-5.2.0-linux64-debug.tar.gz
    ff4174bcdb7068756c33e8103293ddc76e0221a9df50124bf0726975c9e56e1e  zcash-5.2.0-linux64.tar.gz
- in_manifest: |-
    bff3e4f2f76db4f6a04f17f7d8d3fbf61a1e10a75a233f73b1f5b89c2d487968  zcash-5.2.0-rc1-desc.yml
    git:523790613cdefe32c78f61ff80c3ff9c09a893ed zcash
- base_manifests: !!omap
  - bullseye-amd64: |
vagrant@zcash-build:~/gitian.sigs$ 
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 1]
```diff
@@ -0,0 +1,141 @@
+---
+name: "zcash-5.3.0-rc1"
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:2]: Should this be "zcash-5.3.0"?

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+script: |
+  WRAP_DIR=$HOME/wrapped
+  HOSTS="x86_64-linux-gnu"
+  CONFIGFLAGS="--enable-glibc-back-compat --enable-reduce-exports --disable-bench --enable-hardening --enable-werror"
+  MAKEOPTS="V=1 -j$(nproc)"
+  FAKETIME_HOST_PROGS=""
+  FAKETIME_PROGS="date ar ranlib nm strip objcopy"
+  HOST_CFLAGS=""
+  HOST_CXXFLAGS=""
+  HOST_LDFLAGS=-static-libstdc++
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:44]: Why is this needed? I thought it was provided by default.

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+  export QT_RCC_TEST=0
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:46]: This is a QT-specific variable and we no longer use QT.
```suggestion
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  export PATH=${WRAP_DIR}:${PATH}
+
+  # Faketime for depends so intermediate results are comparable
+  create_global_faketime_wrappers "2000-01-01 12:00:00"
+  create_per-host_faketime_wrappers "2000-01-01 12:00:00"
+
+  cd zcash
+  BASEPREFIX=`pwd`/depends
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:89]: There are changes to this file that aren't reflected in `gitian-linux.yml` (e.g., using `pwd` rather than `${PWD}`). Are these necessary for the parallelization, or are they independent changes that should be made in `gitian-linux.yml` as well?

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  # Faketime for binaries
+  create_global_faketime_wrappers "${REFERENCE_DATETIME}"
+  create_per-host_faketime_wrappers "${REFERENCE_DATETIME}"
+
+  # Create the release tarball using (arbitrarily) the first host
+  ./autogen.sh
+  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
+  make dist
+  SOURCEDIST=`echo zcash-*.tar.gz`
+  DISTNAME=`echo ${SOURCEDIST} | sed 's/.tar.*//'`
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:104]: The `.` characters are treated as a regex metachar, not a literal dot:
```shell
$ echo footarbaz | sed 's/.tar.*//'
fo
$ echo footarbaz | sed 's/[.]tar[.].*//'
footarbaz
$ echo footarbaz.tar.gz | sed 's/tar.*//'
foo
$ echo footarbaz.tar.gz | sed 's/[.]tar[.].*//'
footarbaz
```
```suggestion
 DISTNAME=`echo ${SOURCEDIST} | sed 's/[.]tar[.].*//'`
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+  # Faketime for binaries
+  create_global_faketime_wrappers "${REFERENCE_DATETIME}"
+  create_per-host_faketime_wrappers "${REFERENCE_DATETIME}"
+
+  # Create the release tarball using (arbitrarily) the first host
+  ./autogen.sh
+  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
+  make dist
+  SOURCEDIST=`echo zcash-*.tar.gz`
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:103]:
```shell
$ touch zcash-johnny-tables.tar.gz\ injected-arguments\ .tar.gz
$ SOURCEDIST=`echo zcash-*.tar.gz`
$ echo ${SOURCEDIST}
zcash-johnny-tables.tar.gz injected-arguments .tar.gz
$ mkdir -p temp
$ pushd temp
/home/daira/zecc/zcash/temp /home/daira/zecc/zcash
$ touch foo
$ tar cf ../zcash-johnny-tables.tar.gz foo
$ tar xf ../$SOURCEDIST
tar: injected-arguments: Not found in archive
tar: .tar.gz: Not found in archive
tar: Exiting with failure status due to previous errors
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  DISTNAME=`echo ${SOURCEDIST} | sed 's/.tar.*//'`
+  # Correct tar file order
+  mkdir -p temp
+  pushd temp
+  tar xf ../$SOURCEDIST
+  find zcash* | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > ../$SOURCEDIST
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:109]:
```suggestion
  tar xf "../$SOURCEDIST"
  find zcash* | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > "../$SOURCEDIST"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  popd
+
+  ORIGPATH="$PATH"
+  # Extract the release tarball into a dir for each host and build
+  for i in ${HOSTS}; do
+    export PATH=${BASEPREFIX}/${i}/native/bin:${ORIGPATH}
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:115]:
```suggestion
    export PATH="${BASEPREFIX}/${i}/native/bin:${ORIGPATH}"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    mkdir -p distsrc-${i}
+    cd distsrc-${i}
+    INSTALLPATH=`pwd`/installed/${DISTNAME}
+    mkdir -p ${INSTALLPATH}
+    tar --strip-components=1 -xf ../$SOURCEDIST
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:120]:
```suggestion
    export PATH="${BASEPREFIX}/${i}/native/bin:${ORIGPATH}"
    mkdir -p "distsrc-${i}"
    cd "distsrc-${i}"
    INSTALLPATH="`pwd`/installed/${DISTNAME}"
    mkdir -p "${INSTALLPATH}"
    tar --strip-components=1 -xf "../$SOURCEDIST"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    find . -name "lib*.la" -delete
+    find . -name "lib*.a" -delete
+    rm -rf ${DISTNAME}/lib/pkgconfig
+    find ${DISTNAME}/bin -type f -executable -exec objcopy --only-keep-debug {} {}.dbg \; -exec strip -s {} \; -exec objcopy --add-gnu-debuglink={}.dbg {} \;
+    # Commented out while we don't build any libraries
+    #find ${DISTNAME}/lib -type f -exec objcopy --only-keep-debug {} {}.dbg \; -exec strip -s {} \; -exec objcopy --add-gnu-debuglink={}.dbg {} \;
+    find ${DISTNAME} -not -name "*.dbg" | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > ${OUTDIR}/${DISTNAME}-${i}.tar.gz
+    find ${DISTNAME} -name "*.dbg" | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > ${OUTDIR}/${DISTNAME}-${i}-debug.tar.gz
+    cd ../../
+    rm -rf distsrc-${i}
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:136]:
```suggestion
    CONFIG_SITE="${BASEPREFIX}/${i}/share/config.site" ./configure --prefix=/ --disable-ccache --disable-maintainer-mode --disable-dependency-tracking "${CONFIGFLAGS}" CFLAGS="${HOST_CFLAGS}" CXXFLAGS="${HOST_CXXFLAGS}" LDFLAGS="${HOST_LDFLAGS}"
    make ${MAKEOPTS}
    make ${MAKEOPTS} -C src check-security
    make install DESTDIR="${INSTALLPATH}"
    cd installed
    find . -name "lib*.la" -delete
    find . -name "lib*.a" -delete
    rm -rf "${DISTNAME}/lib/pkgconfig"
    find "${DISTNAME}/bin" -type f -executable -exec objcopy --only-keep-debug {} {}.dbg \; -exec strip -s {} \; -exec objcopy --add-gnu-debuglink={}.dbg {} \;
    # Commented out while we don't build any libraries
    #find "${DISTNAME}/lib" -type f -exec objcopy --only-keep-debug {} {}.dbg \; -exec strip -s {} \; -exec objcopy --add-gnu-debuglink={}.dbg {} \;
    find "${DISTNAME}" -not -name "*.dbg" | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > "${OUTDIR}/${DISTNAME}-${i}.tar.gz"
    find "${DISTNAME}" -name "*.dbg" | sort | tar --no-recursion --mode='u+rw,go+r-w,a+X' --owner=0 --group=0 -c -T - | gzip -9n > "${OUTDIR}/${DISTNAME}-${i}-debug.tar.gz"
    cd ../../
    rm -rf "distsrc-${i}"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  done
+  mkdir -p $OUTDIR/src
+  mv $SOURCEDIST $OUTDIR/src
+  mv ${OUTDIR}/${DISTNAME}-x86_64-*-debug.tar.gz ${OUTDIR}/${DISTNAME}-linux64-debug.tar.gz
+  mv ${OUTDIR}/${DISTNAME}-x86_64-*.tar.gz ${OUTDIR}/${DISTNAME}-linux64.tar.gz
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:141]:
```suggestion
  mkdir -p "$OUTDIR/src"
  mv "$SOURCEDIST" "$OUTDIR/src"
  mv "${OUTDIR}/${DISTNAME}-x86_64-*-debug.tar.gz" "${OUTDIR}/${DISTNAME}-linux64-debug.tar.gz"
  mv "${OUTDIR}/${DISTNAME}-x86_64-*.tar.gz" "${OUTDIR}/${DISTNAME}-linux64.tar.gz"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml (hunk 4 above)
Comment (daira) [reply contrib/gitian-descriptors/gitian-linux-parallel.yml:89]:
```suggestion
  BASEPREFIX="`pwd`/depends"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+  # Faketime for depends so intermediate results are comparable
+  create_global_faketime_wrappers "2000-01-01 12:00:00"
+  create_per-host_faketime_wrappers "2000-01-01 12:00:00"
+
+  cd zcash
+  BASEPREFIX=`pwd`/depends
+  # Build dependencies for each host
+  for i in $HOSTS; do
+    make ${MAKEOPTS} -C ${BASEPREFIX} HOST="${i}"
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:92]:
```suggestion
    make ${MAKEOPTS} -C "${BASEPREFIX}" HOST="${i}"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+  done
+
+  # Faketime for binaries
+  create_global_faketime_wrappers "${REFERENCE_DATETIME}"
+  create_per-host_faketime_wrappers "${REFERENCE_DATETIME}"
+
+  # Create the release tarball using (arbitrarily) the first host
+  ./autogen.sh
+  CONFIG_SITE=${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:101]:
```suggestion
  CONFIG_SITE="${BASEPREFIX}/`echo "${HOSTS}" | awk '{print $1;}'`/share/config.site ./configure --prefix=/"
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+        echo "REAL=\`which -a ${i}-${prog} | grep -v ${WRAP_DIR}/${i}-${prog} | head -1\`" >> ${WRAP_DIR}/${i}-${prog}
+        echo 'export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/faketime/libfaketime.so.1' >> ${WRAP_DIR}/${i}-${prog}
+        echo "export FAKETIME=\"$1\"" >> ${WRAP_DIR}/${i}-${prog}
+        echo "\$REAL \$@" >> $WRAP_DIR/${i}-${prog}
+        chmod +x ${WRAP_DIR}/${i}-${prog}
+    done
+  done
+  }
+
+  export PATH=${WRAP_DIR}:${PATH}
```
Comment (daira) [inline contrib/gitian-descriptors/gitian-linux-parallel.yml:82]:
```suggestion
  mkdir -p "${WRAP_DIR}"
  if test -n "$GBUILD_CACHE_ENABLED"; then
    export SOURCES_PATH="${GBUILD_COMMON_CACHE}"
    export BASE_CACHE="${GBUILD_PACKAGE_CACHE}"
    mkdir -p "${BASE_CACHE}" "${SOURCES_PATH}"
  fi

  function create_global_faketime_wrappers {
  for prog in ${FAKETIME_PROGS}; do
    echo '#!/usr/bin/env bash' > "${WRAP_DIR}/${prog}"
    echo "REAL=\`which -a "${prog}" | grep -v "${WRAP_DIR}/${prog}" | head -1\`" >> "${WRAP_DIR}/${prog}"
    echo 'export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/faketime/libfaketime.so.1' >> "${WRAP_DIR}/${prog}"
    echo "export FAKETIME=\"$1\"" >> "${WRAP_DIR}/${prog}"
    echo "\$REAL \$@" >> "$WRAP_DIR/${prog}"
    chmod +x "${WRAP_DIR}/${prog}"
  done
  }

  function create_per-host_faketime_wrappers {
  for i in "$HOSTS"; do
    for prog in ${FAKETIME_HOST_PROGS}; do
        echo '#!/usr/bin/env bash' > "${WRAP_DIR}/${i}-${prog}"
        echo "REAL=\`which -a ${i}-${prog} | grep -v ${WRAP_DIR}/${i}-${prog} | head -1\`" >> "${WRAP_DIR}/${i}-${prog}"
        echo 'export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/faketime/libfaketime.so.1' >> "${WRAP_DIR}/${i}-${prog}"
        echo "export FAKETIME=\"$1\"" >> "${WRAP_DIR}/${i}-${prog}"
        echo "\$REAL \$@" >> "$WRAP_DIR/${i}-${prog}"
        chmod +x "${WRAP_DIR}/${i}-${prog}"
    done
  done
  }

  export PATH="${WRAP_DIR}:${PATH}"
```

**Review by daira — COMMENTED** [review]
I know all of these potential injection vulns were in the previous script, and now might not be the time to fix them. It just bugs me, that's all.

**Comment by superbaud** [comment]
Most of them reflect the same issue as in the extant gitian descriptor file, though there's some that don't, since some improvements were backported into `zcash`/zcash  from bitcoin core after i made the changes, and I didn't check for that after marking my PR ready for review.

To address this, I rebased my PR (the only difference between the non-parallel descriptor and my parallel descriptor is `-j$(nproc)`), and after the 5.3.0 release, like you suggested, we can make a PR that fixes the (now-common) issues with the gitian descriptors that you identified:

```
sasha_work@seasonal-dream gitian-descriptors % cp gitian-linux-parallel.yml ~
sasha_work@seasonal-dream gitian-descriptors % git switch master
Switched to branch 'master'
Your branch is up to date with 'origin-zcash/master'.
sasha_work@seasonal-dream gitian-descriptors % git diff 
sasha_work@seasonal-dream gitian-descriptors % diff ~/gitian-linux-parallel.yml gitian-linux.yml 
2c2
< name: "zcash-5.3.0-rc1"
---
> name: "zcash-5.2.0"
39c39
<   MAKEOPTS="V=1 -j$(nproc)"
---
>   MAKEOPTS="V=1"
sasha_work@seasonal-dream gitian-descriptors % git switch version-5.3.0
Switched to branch 'version-5.3.0'
Your branch is up to date with 'origin-zcash/version-5.3.0'.
sasha_work@seasonal-dream gitian-descriptors % diff ~/gitian-linux-parallel.yml gitian-linux.yml
39c39
<   MAKEOPTS="V=1 -j$(nproc)"
---
>   MAKEOPTS="V=1"
sasha_work@seasonal-dream gitian-descriptors % 
```

File: contrib/gitian-descriptors/gitian-linux-parallel.yml (hunk 1 above)
Comment (superbaud) [reply contrib/gitian-descriptors/gitian-linux-parallel.yml:2]: I was going directly off what the current `gitian-linux.yml` file contained (which included an `rc1`)

File: contrib/gitian-descriptors/gitian-linux-parallel.yml (hunk 2 above)
Comment (superbaud) [reply contrib/gitian-descriptors/gitian-linux-parallel.yml:44]: Indeed same as the `gitian-linux.yml`.

**Review by nuttycom — APPROVED** [review]
utACK

File: contrib/gitian-descriptors/gitian-linux-parallel.yml (hunk 1 above)
Comment (daira) [reply contrib/gitian-descriptors/gitian-linux-parallel.yml:2]: `make-release.py` has a function at https://github.com/zcash/zcash/blob/cc34f4f9bacdaaee64f5d80550229f001959efdb/zcutil/make-release.py#L377 to patch the gitian descriptor file, which should be changed to patch both files (equivalently, pass the filename into the function and call it twice).

Comment (daira) [reply contrib/gitian-descriptors/gitian-linux-parallel.yml:2]: Oh right you already did that :-)

**Review by daira — APPROVED** [review]
utACK
